zebrad/src/commands/start/sync.rs

```diff
@@ -198,97 +197,64 @@ where
+                .chain(self.prospective_tips.iter().cloned())
+                .collect(),
+        )
+        .await?;
 
         Ok(())
     }
 
     /// Queue downloads for each block that isn't currently known to our node
-    async fn request_blocks(&mut self, mut hashes: Vec<BlockHeaderHash>) -> Result<(), Report> {
-        hashes.retain(|hash| !self.known_block(hash));
-
+    async fn request_blocks(&mut self, hashes: Vec<BlockHeaderHash>) -> Result<(), Report> {
         for chunk in hashes.chunks(10usize) {
-            self.queue_download(chunk).await?;
-        }
+            let set = chunk.iter().cloned().collect();
 
-        Ok(())
-    }
+            let request = self
+                .peer_set
+                .ready_and()
+                .await
+                .map_err(|e| eyre!(e))?
+                .call(zn::Request::BlocksByHash(set));
 
-    /// Drive block downloading futures to completion and dispatch downloaded
-    /// blocks to the validator
-    async fn process_blocks(&mut self) -> Result<(), Report> {
-        info!(in_flight = self.block_requests.len(), "processing blocks");
+            let mut verifier = self.verifier.clone();
 
-        while let Some(res) = self.block_requests.next().await {
-            match res.map_err::<Report, _>(|e| eyre!(e)) {
-                Ok(zn::Response::Blocks(blocks)) => {
-                    info!(count = blocks.len(), "received blocks");
-                    for block in blocks {
-                        let hash = block.as_ref().into();
-                        assert!(
-                            self.downloading.remove(&hash),
-                            "all received blocks should be explicitly requested and received once"
-                        );
-                        let _ = self.downloaded.insert(hash);
-                        self.validate_block(block).await?;
-                    }
-                }
-                Ok(_) => continue,
-                Err(e) => {
-                    error!("{:?}", e);
-                }
-            }
-        }
+            let _ = tokio::spawn(async move {
+                match async move {
+                    let resp = request.await?;
 
-        Ok(())
-    }
+                    if let zn::Response::Blocks(blocks) = resp {
+                        debug!(count = blocks.len(), "received blocks");
 
-    /// Validate a downloaded block using the validator service, inserting the
-    /// block into the state if successful
-    #[tracing::instrument(skip(self))]
-    async fn validate_block(
-        &mut self,
-        block: std::sync::Arc<zebra_chain::block::Block>,
-    ) -> Result<(), Report> {
-        let fut = self
-            .state
-            .ready_and()
-            .await
-            .map_err(|e| eyre!(e))?
-            .call(zs::Request::AddBlock { block });
+                        for block in blocks {
+                            let _hash = verifier.ready_and().await?.call(block).await?;
```

Comment:
I think this will potentially have deadlocks because the verifications aren't independently driven tasks, but changing it could be a separate chunk of work to this PR.

Reply:
https://github.com/ZcashFoundation/zebra/pull/537#discussion_r445125658

Reply:
I swear I fixed this but I cant find where, gah

Reply:
must have been in one of the other PRs that got closed

Reply:
In particular, the checkpoint verifier makes all the blocks between a pair of checkpoints depend on each other.

So no blocks complete until we have the entire chain between the checkpoints.

For details, see https://github.com/ZcashFoundation/zebra/pull/537#discussion_r447383033